accepted/00000-salt-ssh-with-salt-bundle.md

```diff
@@ -0,0 +1,77 @@
+
+This approach will help us to limit the number of supported versions of Salt codebase and drop `py26-compat-salt`, `py27-compat-salt` (as these packages are shipping `2016.11.10` and `3000.3` respectively) and all Python 2 dependencies maintenance.
+So we don't need to maintain old salt codebases, but focus only on the latest one.
+
+# Detailed design
+[design]: #detailed-design
+
+## Uyuni roster
+
+We can use Uyuni roster module to provide roster data instead of creating flat roster files with Java. PR moving the logic to Uyuni roster on Java side: https://github.com/uyuni-project/uyuni/pull/4457
+
+This change can help to avoid creating temporary roster files (https://docs.saltproject.io/en/latest/topics/ssh/roster.html) and implements caching additionally.
+
+## Pre flight script
+
+Salt can execute a shell script on the salt-ssh client before the "real salt-ssh execution" starts.
+This script downloads the salt-bundle package from the appropriate bootstrap repository and extracts the virtual environment.
+By-default there is no way to pass additional parameters to pre fligt script running on the managed system.
+
+The "Pre flight script" calculates the bootstrap repository URL in a similar manner to the existing boostrap.sh script, based on the operating system and OS architecture of the system it is executed on.
+It requires one parameter from the Uyuni server: (host, port) (entry point) for SSH connections through tunnels.
+The salt codebase need to be modified to pass additional parameters, otherwise we have to generate pre flight script individually for each managed system.
+
+## Deploying Salt Bundle to the managed system
+
+We don't need to install Salt Bundle package on the managed system.
+We need to extract the virtual environment of the bundle from the package.
+Pre flight script is deploying it as described above.
+
+## Using proper salt codebase
+
+On deploying Salt Bundle with pre flight script, without modification salt-thin (`salt-thin` is a tarball with salt codebase created by the server to deploy salt on salt-ssh managed system: https://docs.saltproject.io/en/latest/ref/runners/all/salt.runners.thin.html) deployment, we will get two different codebases of salt deployed on the managed system (one codebase is from salt bundle package for the OS+arch and the other one is from salt-master delivered with salt-thin tarball).
+As the sources of these two codebases are different (update channel of the Server and client tools relevant for the system) these codebases are not in sync.
+We can prevent deploying salt-thin in case if Salt Bundle has been deployed already and use it as preferred salt codebase to handle salt-ssh session with on managed system side.
+In most cases the client tools channels are updating more frequently than the server and for the minions (not salt-ssh systems) we are using the salt codebase from client tools, it's better to use Salt Bundle codebase for salt-ssh systems.
```

Comment:
IIUC this means we do not expect to ever use the salt-thin any longer. Is that correct, or are there exceptions?

Reply:
With the described approach and in my testing environment `salt-thin` is deployed only in case if salt bundle was not deployed for some reason (bootstrap repo is unavalaible or doesn't contain the bundle, or http(s) from the server is not accessible) and in case of using `salt-thin` python 3 must be installed on the system